File: multiplayer/packages/kernel/src/logic/canonical-json.ts
/**
 * Deterministic JSON, the form every order digest is taken over: object keys sorted by UTF-16 code
 * unit (JavaScript's default string order, .NET's `StringComparer.Ordinal`), no whitespace, and
 * numbers restricted to safe integers.
 *
 * The restriction is the point: a client written in another language has to reproduce this text
 * byte for byte to verify `ordersHash`, and floating-point shortest round-trip text is not portable
 * (`1e+21` here, `1E+21` in .NET). Integers, booleans, null and JSON-escaped strings are. Anything
 * outside that is a programming error, not a runtime condition, so it throws rather than hashing
 * something a peer cannot reproduce.
 */
export function canonicalJson(value: unknown): string {
  return JSON.stringify(canonicalize(value))
}

export class NonCanonicalValueError extends Error {
  constructor(
    readonly path: string,
    detail: string,
  ) {
    super(`cannot canonicalize ${path || 'value'}: ${detail}`)
    this.name = 'NonCanonicalValueError'
  }
}

type PathSegment = string | number

function canonicalize(value: unknown, path: PathSegment[] = []): unknown {
  if (value === null || typeof value === 'boolean' || typeof value === 'string') return value
  if (typeof value === 'number') {
    if (!Number.isSafeInteger(value) || Object.is(value, -0)) {
      throw new NonCanonicalValueError(
        describe(path),
        `${value} is not a safe non-negative-zero integer`,
      )
    }
    return value
  }
  if (Array.isArray(value)) return value.map((item, index) => descend(path, index, item))
  if (typeof value === 'object') {
    const source = value as Record<string, unknown>
    const sorted: Record<string, unknown> = {}
    for (const key of Object.keys(source).sort()) {
      sorted[key] = descend(path, key, source[key])
    }
    return sorted
  }
  throw new NonCanonicalValueError(describe(path), `unsupported type ${typeof value}`)
}

function descend(path: PathSegment[], segment: PathSegment, value: unknown): unknown {
  path.push(segment)
  try {
    return canonicalize(value, path)
  } finally {
    path.pop()
  }
}

/**
 * Renders the accumulated segments the way the C# mirror's `path` parameter spells them
 * (`CanonicalJson.WriteObject`/`WriteArray`): an index is always `[n]`, and a key is dotted unless
 * nothing has been rendered yet. The emptiness of what came before decides the dot, not the
 * segment's position -- an empty-string key renders as nothing, so `{ '': { y: 1.5 } }` reports
 * `y` rather than `.y`, matching both the mirror and the string-building version this replaced.
 *
 * Only a throw reaches here, so the allocation is off the hot path.
 */
function describe(path: readonly PathSegment[]): string {
  let rendered = ''
  for (const segment of path) {
    if (typeof segment === 'number') rendered += `[${segment}]`
    else rendered += rendered.length === 0 ? segment : `.${segment}`
  }
  return rendered
}

File: src/Rechaos.Multiplayer/Protocol/CanonicalJson.cs
using System.Globalization;
using System.Text;
using System.Text.Json;

namespace Rechaos.Multiplayer.Protocol;

/// <summary>
/// Deterministic JSON: the exact text the server's order digests are taken over.
/// </summary>
/// <remarks>
/// <para>
/// The rules are the server's, in <c>packages/kernel/src/logic/canonical-json.ts</c>: object keys
/// sorted by UTF-16 code unit (which is <see cref="StringComparer.Ordinal"/>), no whitespace,
/// standard JSON string escaping, and integers written plainly.
/// </para>
/// <para>
/// The restriction to integers is the point. A client in another language has to reproduce this
/// text byte for byte to verify a digest, and a floating-point value's shortest round-trip
/// spelling is not portable: JavaScript writes <c>1e+21</c> where .NET writes <c>1E+21</c>. Any
/// value outside the portable set throws rather than hashing something a peer cannot reproduce.
/// </para>
/// </remarks>
public static class CanonicalJson
{
    /// <summary>The canonical text of a value already parsed as JSON.</summary>
    public static string Encode(JsonElement value)
    {
        var builder = new StringBuilder();
        Write(builder, value, string.Empty);
        return builder.ToString();
    }

    /// <summary>The canonical text as UTF-8 bytes, which is what a digest is taken over.</summary>
    public static byte[] EncodeUtf8(JsonElement value) => Encoding.UTF8.GetBytes(Encode(value));

    private static void Write(StringBuilder builder, JsonElement value, string path)
    {
        switch (value.ValueKind)
        {
            case JsonValueKind.Object:
                WriteObject(builder, value, path);
                return;
            case JsonValueKind.Array:
                WriteArray(builder, value, path);
                return;
            case JsonValueKind.String:
                WriteString(builder, value.GetString() ?? string.Empty);
                return;
            case JsonValueKind.Number:
                WriteNumber(builder, value, path);
                return;
            case JsonValueKind.True:
                builder.Append("true");
                return;
            case JsonValueKind.False:
                builder.Append("false");
                return;
            case JsonValueKind.Null:
                builder.Append("null");
                return;
            case JsonValueKind.Undefined:
            default:
                throw new NonCanonicalValueException(path, $"unsupported kind {value.ValueKind}");
        }
    }

    private static void WriteObject(StringBuilder builder, JsonElement value, string path)
    {
        var properties = new List<JsonProperty>();
        foreach (var property in value.EnumerateObject()) properties.Add(property);
        properties.Sort(static (left, right) => string.CompareOrdinal(left.Name, right.Name));
        builder.Append('{');
        for (var index = 0; index < properties.Count; index++)
        {
            if (index > 0) builder.Append(',');
            var property = properties[index];
            WriteString(builder, property.Name);
            builder.Append(':');
            Write(builder, property.Value, path.Length == 0 ? property.Name : $"{path}.{property.Name}");
        }
        builder.Append('}');
    }

    private static void WriteArray(StringBuilder builder, JsonElement value, string path)
    {
        builder.Append('[');
        var index = 0;
        foreach (var item in value.EnumerateArray())
        {
            if (index > 0) builder.Append(',');
            Write(builder, item, $"{path}[{index}]");
            index++;
        }
        builder.Append(']');
    }

    /// <summary>
    /// Numbers are safe integers or nothing.
    /// </summary>
    /// <remarks>
    /// The raw text is re-parsed rather than reprinted from the reader's own formatting so that
    /// <c>1.0</c>, <c>1e3</c> and <c>-0</c> are all refused: each of them is a number some writer
    /// spells differently, and the digest is over the spelling.
    /// </remarks>
    private static void WriteNumber(StringBuilder builder, JsonElement value, string path)
    {
        var raw = value.GetRawText();
        if (!long.TryParse(raw, NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture, out var integer)
            || raw.StartsWith("-0", StringComparison.Ordinal) && integer == 0)
        {
            throw new NonCanonicalValueException(
                path, $"{raw} is not an integer with one portable spelling");
        }
        if (integer is > MaxSafeInteger or < -MaxSafeInteger)
        {
            throw new NonCanonicalValueException(path, $"{raw} is outside the safe integer range");
        }
        builder.Append(integer.ToString(CultureInfo.InvariantCulture));
    }

    /// <summary>
    /// <c>JSON.stringify</c>'s string escaping, which is narrower than what .NET writes by default.
    /// </summary>
    /// <remarks>
    /// Only the quote, the backslash and the control characters below U+0020 are escaped, and the
    /// five short forms are preferred where they exist. Everything else, non-ASCII included, is
    /// written literally: escaping more would still be valid JSON and a different byte sequence,
    /// which is a different digest.
    /// </remarks>
    private static void WriteString(StringBuilder builder, string value)
    {
        builder.Append('"');
        foreach (var character in value)
        {
            switch (character)
            {
                case '"': builder.Append("\\\""); break;
                case '\\': builder.Append("\\\\"); break;
                case '\b': builder.Append("\\b"); break;
                case '\f': builder.Append("\\f"); break;
                case '\n': builder.Append("\\n"); break;
                case '\r': builder.Append("\\r"); break;
                case '\t': builder.Append("\\t"); break;
                default:
                    if (character < ' ') builder.Append(CultureInfo.InvariantCulture, $"\\u{(int)character:x4}");
                    else builder.Append(character);
                    break;
            }
        }
        builder.Append('"');
    }

    /// <summary>2^53 - 1: the largest integer JavaScript holds exactly, and so the wire's ceiling.</summary>
    private const long MaxSafeInteger = 9_007_199_254_740_991;
}

/// <summary>A value with no portable JSON text, so no digest a peer could reproduce.</summary>
public sealed class NonCanonicalValueException(string path, string detail)
    : InvalidOperationException($"cannot canonicalize {(path.Length == 0 ? "value" : path)}: {detail}")
{
    /// <summary>Where in the document the offending value sits, e.g. <c>ops[0].gang</c>.</summary>
    public string Path { get; } = path;
}
